Fix issue with subset-based peakGroups alignment and other updates

jorainer opened the pull request:
This fixes issue #702 and in addition splits the peakGroups alignment function into two parts (finally) supporting also the use of externally defined alignment matrices.

jorainer committed:
comments: some comments and small refactoring

jorainer committed:
fix: adapt to Spectra 1.13.2 with chunk-wise processing

- Ensure to disable chunk-wise processing for all peaksData calls by passing
  parameter f = factor().

jorainer committed:
feat: add filterMsLevel method for MsExperiment and XcmsExperiment

- Add the `filterMsLevel` methods for `MsExperiment` and `XcmsExperiment`.

jorainer committed:
refactor: split PeakGroups alignment function

- Split the PeakGroups alignment function to enable alignment using a predefined
  peak matrix.

jorainer committed:
fix: minFraction not correctly used for subset-based peakGroups

- Fix issue #702 to ensure peakGroups subset-based alignment correctly uses the
  `minFraction` parameter.
- Split the code for peakGroups-based alignment to enable alignment using
  externally defined retention time matrix of anchor peaks.

jorainer committed:
docs: fix documentation

jorainer committed:
Merge branch 'devel' into jomain

jorainer committed:
update: Update Spectra objects

jorainer commented:
Note: unit tests fail until Spectra is updates.

jorainer committed:
refactor: clean up code

jorainer committed:
fix: missed passing of span parameter in new function

jorainer committed:
fix: ensure ms_level column in chromPeaksData is integer

- Ensure the "ms_level" column in the `chromPeaksData` is of type `integer` to
  avoid obscure validity errors on extracted `Chromatogram` objects (happening
  randomly).

jorainer committed:
refactor: allow peakGroups alignment using prefefined anchor matrix

- Support alignment with the peakGroups method using a predefined anchor matrix.

jorainer committed:
docs: update NEWS

philouail commented on R/XcmsExperiment.R:
```diff
@@ -1332,35 +1351,28 @@ setMethod(
+            subset = subset(param), subsetAdjust = subsetAdjust(param))
         pt <- vapply(object@processHistory, processType, character(1))
         idx_pg <- .match_last(.PROCSTEP.PEAK.GROUPING, pt, nomatch = -1L)
         if (idx_pg > 0)
```
the double entry has to happen here I guess @jorainer

jorainer replied:
actually, it was the `.match_last` call. Instead of returning `-1L` if no peak grouping was performed it returned a positive integer (I think it was 3 in that particular case), so, subset the processHistory with that "out of bound" index returned/added the `NULL` entry. Stupid but oin the `.match_last` function.

jorainer committed:
fix: fix the internal .match_last function

- `.match_last` was returning a wrong result if used with `nomatch = -1`. This
  affected removing or adding elements to the processHistory during xcms
  preprocessing.

jorainer committed:
docs: rename variable in vignette (issue #704)

jorainer committed:
refactor: throw error for adjustRtime with alignment results

- Throw an error for `adjustRtime` if alignment results are already present.

jorainer committed:
ci: use BioC 3.19 for macOS and Windows

jorainer committed:
Merge branch 'devel' into jomain

jorainer committed:
ci: use R "devel" version

jorainer committed:
refactor: fix C format complains for printf

jorainer committed:
fix: use %llu as format in sprintf C calls

jorainer committed:
fix: fix format for integer in C code

jorainer committed:
fix: remove printing of numbers for C errors

jorainer commented:
OK. unit tests fixed - ready for review/comments @sneumann

sneumann commented on R/MsExperiment.R:
```diff
@@ -20,6 +20,14 @@ setMethod("filterMz", "MsExperiment",
 
+#' @rdname XcmsExperiment
+setMethod("filterMsLevel", "MsExperiment",
+          function(object, msLevel. = uniqueMsLevels(object)) {
```
First time I see a parameter with trailing dot. Is there a pattern I need to learn ?

jorainer replied:
Yes, that's rather unusual, I agree. It's still from *MSnbase* days, without the `.` R was complaining about recursive arguments - I guess that came from the R dispatcher that got confused by a parameter `msLevel` and a `msLevel` method. For consistency and backward compatibility we used this also in other method definitions in `Spectra`. Thanks to the partial parameter matching from R users can use `msLevel = 1L` as well as `msLevel. = 1L`.

I know it looks ugly, but we can't change it here, because that comes from far upstream...

sneumann commented on R/do_adjustRtime-functions.R:
```diff
@@ -115,282 +121,184 @@ do_adjustRtime_peakGroups <-
-                " than 2x. This is dangerous and the algorithm is probably ",
-                "overcorrecting your data. Consider increasing the span ",
-                "parameter or switching to the linear smoothing method.")
+        rtime_adj
```
I like a more explicit return(rtime_adj), unless we use a style convention that suggests brevity rulez

jorainer replied:
I also used `return()` a lot - but then had a chat with Martin (Morgan) and he mentioned that for R there's no need to have yet another function call at the end of a function since the last variable will anyway be returned by default. `return` should only be used e.g. to break within a function and still return something. That seems to be now the R coding style.

sneumann commented on R/do_adjustRtime-functions.R:
```diff
@@ -485,6 +393,10 @@ do_adjustRtime_peakGroups_orig <- function(peaks, peakIndex, rtime,
 #' @details This function is called internally by the
 #'     do_adjustRtime_peakGroups function and the retcor.peakgroups method.
 #'
+#' Update for version 4.1.2: correctly consider the `sampleIndex` and
```
I think 4.1.4

jorainer replied:
yes sure, thanks! will fix.

sneumann commented on data/faahko_sub2.RData:
Shouldn't that be in faahKO ?

jorainer replied:
Good question. This basically is "just" a small object to have simpler examples (i.e. load the data without having to first perform all preprocessing to then just run the example). 

IMHO the faahKO should "only" contain the raw data and a object of a class from `xcms` should be in `xcms` - that way `faahKO` is independent of `xcms` and we avoid eventual cyclic redundancies.

sneumann commented on data/xmse.RData:
And this would go into msdata ?

jorainer replied:
Same here, IMHO `msdata` should contain raw data files and be independent of data structures.

sneumann commented on tests/testthat/test_Chromatogram.R:
```diff
@@ -1,5 +1,5 @@
 ## test_that("extractChromatograms is deprecated", {
```
Not only deprecated, but gone. So test can go out of the window as well

jorainer replied:
yep - will remove.

sneumann reviewed (commented):
Mostly cosmetics

jorainer committed:
refactor: fix Steffen's comments
